README.md
pyshmup
=======

Schmup sample using pygame, Box2D and BulletML

python-bulletml-2/NEWS.txt
2 - 2010.04.24
 * Easier to extend Action with new behaviors.
 * Conditional <if>/<then>/<else> elements.
 * Faster Pyrex-less collision extension.

1 - 2010.04.10
 * Initial release.


test_shoot.py
import os
import sys 

sys.path.append(os.path.abspath(
    os.path.join(os.path.dirname(__file__), 'bloodyhell')
))

from sandbox import SandBox

def run():

    from bloodyhell.game import Game

    game = Game(
        'test shoot',
        (800, 600),
        os.path.join(os.path.dirname(__file__), 'res'),
        fps=24
    )
    navigator = game.navigator()
    navigator.set_current_view(SandBox())
    game.run()

if __name__ == '__main__':
    run()


test_bulletml.py
import os
import sys

import pygame

import bulletml
import bulletml.bulletyaml
from bulletml.collision import collides_all

pygame.display.init()
screen = pygame.display.set_mode([600, 600], pygame.DOUBLEBUF)

filename = os.path.join(os.path.dirname(__file__), 'python-bulletml-2', 'examples', 'boss', '[Guwange]_round_4_boss_eye_ball.xml')
target = bulletml.Bullet()
doc = bulletml.BulletML.FromDocument(open(filename, "rU"))
source = bulletml.Bullet.FromDocument(
    doc, x=150, y=150, target=target, rank=0.5)
                        

red = pygame.Surface([3, 3])
red.fill([255, 0, 0])
green = pygame.Surface([3, 3])
green.fill([0, 255, 0])
blue = pygame.Surface([3, 3])
blue.fill([0, 0, 255])
clock = pygame.time.Clock()
target = bulletml.Bullet()

bullets = dict(red=red, green=green, blue=blue)

active = set([source])
print "lACTIVE"

source.vanished = True

while (True):
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            raise SystemExit
    lactive = list(active)
    for obj in lactive:
        new = obj.step()
        active.update(new)
        if (obj.finished
            or not (-50 < obj.x < 350)
            or not (-50 < obj.y < 350)):
            active.remove(obj)
    if lactive:
        collides = collides_all(target, lactive)
    screen.fill([0, 0, 64] if collides else [0, 0, 0] )
    for obj in active:
        try:
            x, y = obj.x, obj.y
        except AttributeError:
            pass
        else:
            if not obj.vanished:
                x *= 2
                y *= 2
                x -= 1
                y -= 1
                bullet = bullets.get(obj.appearance, red)
                screen.blit(bullet, [x, 600 - y])

    pygame.display.flip()


ship.py
import os
import sys 
from bloodyhell.world.actor import Actor
from bulletpattern import BulletPattern
from bullet import Bullet


class Ship(Actor):

    def __init__(self, position, size, level=None):
        super(Ship, self).__init__(
            'ship', 'ship', position, size
        )
        self._speed = 4.0

        self._x_vel = 0.0
        self._y_vel = 0.0
        self._bullets = []

        self.listen_key('right')
        self.listen_key('left')
        self.listen_key('up')
        self.listen_key('down')
        self.listen_key('space')
        self._level = level

    def update(self):
        super(Ship, self).update()
        self.set_y_velocity(self._y_vel)
        self.set_x_velocity(self._x_vel)

    def bullets(self):
        return self._bullets

    def on_right_pressed(self):
        self._x_vel = self._speed

    def on_right_released(self):
        self._x_vel = 0.0

    def on_left_pressed(self):
        self._x_vel = -self._speed

    def on_left_released(self):
        self._x_vel = 0.0

    def on_up_pressed(self):
        self._y_vel = self._speed

    def on_up_released(self):
        self._y_vel = 0.0

    def on_down_pressed(self):
        self._y_vel = -self._speed

    def on_down_released(self):
        self._y_vel = 0.0
    
    def on_space_pressed(self):
        (x,y) = self.position()
        # self._bullets.append(Bullet((x+1.0, y)))
        path = os.path.join(os.path.dirname(__file__), 'python-bulletml-2', 'examples', 'boss', '[Psyvariar]_X-A_boss_opening.xml')
        self._bullets.append(BulletPattern((x+1.0, y), path, self._level))

    def on_space_released(self):
        pass
        
    def on_collision(self, chunk, point):
        pass


bullet.py
from math import sin, cos
import bulletml
import bulletml.bulletyaml
from bulletml.collision import collides_all

from bloodyhell.world.actor import Actor

class Bullet(Actor):
    def __init__(self, position, bullet, level, size=(0.3, 0.3)):
        super(Bullet, self).__init__(
            'bullet', 'bullet', position, size
        )
        self._level = level
        self._bullet = bullet
        self._position = position
        level.add_chunk(self, level.SPRITES)

    def update(self):
        super(Bullet, self).update()
        # TODO : Step()
        # Replace with speed from BulletML

        liste = self._bullet.step()
        
        for bul in liste:
            if bul != self._bullet:
                Bullet(self._position, bullet = bul, level = self._level)
            else:         
                self._bullet = bul

        self.set_x_velocity(cos(self._bullet.direction)*2)
        self.set_y_velocity(sin(self._bullet.direction)*2)

    def on_collision(self, chunk, point):
        print "bullet collided"

bulletpattern.py
import bulletml
import bulletml.bulletyaml
from bulletml.collision import collides_all
from bullet import Bullet

class BulletPattern():
    def __init__(self, start_pos, path, level):
        target = bulletml.Bullet()
        self._bulletml = bulletml.BulletML.FromDocument(open(path, "rU"))
        self._source = bulletml.Bullet.FromDocument(
            self._bulletml, x=150, y=150, target=target, rank=0.5)
        self._bullets = []
        self._active = set([self._source])
    
        for bul in list(self._active):
            bullet = Bullet(start_pos, bullet=bul, level = level)
            self._bullets.append(bullet)

        print self._active.__len__()

    def update(self, bullet):
        self._active.update(bullet)

    def active(self):
        return self._active

    def remove(self,bullet):
        self._bullets.remove(bullet)

sandbox.py
import os
import sys 

import bulletml
import bulletml.bulletyaml
from bulletml.collision import collides_all

from bloodyhell.game import Game
from bloodyhell.level import Level
from bloodyhell.layer.rect import Rect
from bloodyhell.layer import Layer
from bloodyhell.world.actor import Actor
from bloodyhell.world.fence import Fence
from bloodyhell.world.decoration import Decoration

from ship import Ship

class SandBox(Level):

    def __init__(self):
        RESOLUTION = (800, 600)
        res_width, res_height = RESOLUTION
        super(SandBox, self).__init__(camera_config={
            'target': (5.0, 3.5),
            'width': 15.0,
            'rect': Rect((10, 10), (res_width - 20, res_height - 20)),
            'limits': {'left': 0.0, 'bottom': 0.0,
                       'right': 70.0, 'top': 10.0}
        }, gravity=(0, 0))
        self.listen('quit')
        
        # Add background (filled with skyblue)
        self.add_layer(
            Layer(position=(0, 0), size=RESOLUTION).fill('005656'),
            self.BACKGROUND
        )

        self.loader().load_package('static')
        j1 = 1
        j2 = 5
        for i in range(50):
                self.add_chunk(
                    Fence(
                        (i * 0.49, 2.0 + j1),
                        (0.5, 0.5),
                        'static.brick'
                    ),
                    self.PLATFORM
                )
                self.add_chunk(
                    Fence(
                        (i * 0.49, 2.0 + j2),
                        (0.5, 0.5),
                        'static.brick'
                    ),
                    self.PLATFORM
                )

        # BulletML test
        target = bulletml.Bullet()
        path = os.path.join(os.path.dirname(__file__), 'python-bulletml-2', 'examples', 'normal', 'threefire.xml')
        self._bulletml = bulletml.BulletML.FromDocument(open(path, "rU"))
        source = bulletml.Bullet.FromDocument(
            self._bulletml, x=150, y=150, target=target, rank=0.5)
        self._bullets = [source]
   

        self.loader().load_package('ship')
        self.loader().load_package('bullet')

        self._ship = Ship(position=(1.5, 4.0), size=(1.0, 0.5), level=self)
        self.add_chunk(self._ship, self.SPRITES)

        self.world().camera().watch(self._ship)

    def on_frame(self, delta):
        super(SandBox, self).on_frame(delta)
        # while self._ship.bullets().__len__() > 0:
        #     bullet = self._ship.bullets().pop()

        #     self.add_chunk(bullet, self.SPRITES)
    def on_quit(self, event):
        sys.exit()
